### PR TITLE
Rewrite TPU forwarder test

### LINKED ISSUE #2307: CI unstable test: test_tpu_forwarder
#### Problem

Fails consistently on macos

#### Proposed Solution

Ignore for now

### PR DESCRIPTION
**Pull request by sagar-solana**
#### Problem
The existing TPU forwarder test was unstable when run in parallel with the other tests. Something to do with the timing of messages being delivered to the test process' sockets.
#### Summary of Changes
Minor refactor of the TPU forwarder to pull out some offline functionality. Added a unit test on that offline logic and removed the earlier test.

Fixes #2307

### CODE REVIEW

File: src/tpu_forwarder.rs [hunk 1]
```diff
@@ -105,96 +120,51 @@ impl Service for TpuForwarder {
...
+    pub fn test_update_addrs() {
+        let keypair = Keypair::new();
+        let my_id = keypair.pubkey();
+        // test with no pubkey
+        assert!(!TpuForwarder::update_addrs(
+            None,
+            &my_id,
+            &Arc::new(RwLock::new(Packets::default()))
+        ));
+        // test with no tvu
```
Comment (garious) [inline src/tpu_forwarder.rs]: s/tvu/tpu/?

File: src/tpu_forwarder.rs [hunk 2]
```diff
@@ -36,24 +38,37 @@ impl TpuForwarder {
...
+                .cloned();
+            if TpuForwarder::update_addrs(leader_data, &my_id, &msgs.clone()) {
+                msgs.read().unwrap().send_to(&socket)?
+            }
+            continue;
+        }
+    }
+
+    fn update_addrs(
+        leader_data: Option<ContactInfo>,
```
Comment (garious) [inline src/tpu_forwarder.rs]: Why pass in `Option<ContactInfo>` and not `tpu_addr: &SocketAddr`?

File: src/tpu_forwarder.rs [hunk 3]
```diff
@@ -36,24 +38,37 @@ impl TpuForwarder {
...
+        my_id: &Pubkey,
+        msgs: &Arc<RwLock<Packets>>,
+    ) -> bool {
```
Comment (garious) [inline src/tpu_forwarder.rs]: That returned bool seems arbitrary. Is it an error case?

**Comment by jackcmay** [comment]
Not testing any scenarios that result in failure?

File: src/tpu_forwarder.rs (hunk 1 above)
Comment (sagar-solana) [reply src/tpu_forwarder.rs]: tpu. oops

File: src/tpu_forwarder.rs (hunk 3 above)
Comment (sagar-solana) [reply src/tpu_forwarder.rs]: Yeah, not very rust like. I wasn't a 100% on which way to go between result and bool and it seemed pretty minor so I stuck with bool. I'll update it to a result.

**Comment by sagar-solana** [comment]
> Not testing any scenarios that result in failure?

I'm testing 3 fails and 1 success. Should have all branches of the func covered.

**Pull request merged**